README.md
<a name="readme-top"></a>
[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]
[![LinkedIn][linkedin-shield]][linkedin-url]

<!-- PROJECT LOGO -->
<br />
<div align="center">
  <a href="https://github.com/etherCorps/SK-Redis-SessionManager">
    <img src="https://sveltekit-redis-session-manager.vercel.app/logo.svg" alt="Logo" width="80" height="80">
  </a>

<h3 align="center">SvelteKit Redis Session Manager</h3>

  <p align="center">
    Redis integration in SvelteKit for  Session Management
    <br />
    <a href="https://www.npmjs.com/package/@ethercorps/sveltekit-redis-session">@ethercorps/sveltekit-redis-session</a>
    <br /> 
    <a href="https://dev.to/theether0/redis-integration-in-sveltekit-a-game-changer-for-session-management-84i"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://sveltekit-redis-session-manager.vercel.app">View Demo</a>
    ·
    <a href="https://github.com/etherCorps/SK-Redis-SessionManager/issues">Report Bug</a>
    ·
    <a href="https://github.com/etherCorps/SK-Redis-SessionManager/issues">Request Feature</a>
  </p>
</div>
<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#setup">Installation</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
    <li><a href="#acknowledgments">Acknowledgments</a></li>
  </ol>
</details>

<!-- ABOUT THE PROJECT -->

## About The Project

[//]: # '[![Product Name Screen Shot][product-screenshot]](https://github.com/etherCorps/SK-Redis-SessionManager)'

"sveltekit-redis-session" makes it easy for developers to use Redis as a session manager in SvelteKit
applications. It offers a variety of convenient features that make managing user sessions a breeze. This includes simple
functions for storing and retrieving data, encryption of session data for added security, and automatic handling of
session expiration. Additionally, the package is highly customizable, which allows developers to adjust it to their
specific requirements and optimize performance (COPY MY CODE BUT LEARN).
Overall, "sveltekit-redis-session" is choice for managing user sessions in any SvelteKit application thanks to SvelteKit package for robust functionality and flexibility.

<p align="right">(<a href="#readme-top">back to top</a>)</p>

### Built With

- [![SvelteKit][SvelteKit]][SvelteKit-url]
- [![Redis][Redis]][Redis-url]
- [![Svelte][Svelte.dev]][Svelte-url]

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- GETTING STARTED -->

## Getting Started

This is an example of how you may give instructions on setting up your project locally.
To get a local copy up and running follow these simple example steps.

### Prerequisites

This is an example of how to list things you need to use the software and how to install them.

- Install the sveltekit-redis-session
  ```sh
  pnpm i @ethercorps/SvelteKit-redis-session
  ```

### Setup

<a href="https://dev.to/theether0/redis-integration-in-sveltekit-a-game-changer-for-session-management-84i">Guide to how
to use @ethercorps/SvelteKit-redis-session</a>

1. First we need to make instance of it to use everywhere in project.
   ```ts
   import { RedisSessionStore } from '@ethercorps/SvelteKit-redis-session';
   import Redis from 'ioredis';
   export const sessionManager = new RedisSessionStore({
   	redisClient: new Redis(), // Required A pre-initiated redis client
   	secret: 'your-secret-key', // Required A secret key for encryption and other things,
   	cookieName: 'session', // CookieName to be saved in cookies for browser Default session
   	prefix: 'sk-session', // Prefix for Redis Keys Default sk-session
   	signed: true, // Do you want to sign your cookies Default true
   	encrypted: false, // Do you want to encrypt your cookies using 'aes-256-cbc' algorithm Default false
   	useTTL: true, // Do you wanna use redis's Expire key functionality Default false
   	renewSessionBeforeExpire: false, // Do you wanna update session expire time in built function Default false
   	renewBeforeSeconds: 30 * 60, // If renewSessionBeforeExpire is true define your renew before time in seconds Default 30 minutes
   	serializer: JSON, // You can define your own serializer functions to stringify and parse sessionData for redis Default JSON
   	cookiesOptions: {
   		path: '/',
   		httpOnly: true,
   		sameSite: 'strict',
   		secure: !dev, // From SvelteKit "$app/environment"
   		maxAge: 60 * 60 * 24 // You have to define time in seconds and it's also used for redis key expiry time
   	} // You have more options these are default used in package for more check sveltekit CookieSerializeOptions type.
   });
   ```
   > These are the default config example you can use as per your need and make it better for your use.
   > I have written an article to explain more about this package <a href="">link for article</a>.
2. To create new session and add cookies for user after authentication

   ```ts
   // Example it's a +page.server.ts
   import sessionManager from 'sessionManagerFile';

   export const actions: Actions = {
   	login: async ({ req, cookies, locals }) => {
   		const formdata = await request.formData();
   		// Form validation && user validation
   		const { data, error, message } = sessionManager.createNewSession(cookies, userData);
   		// data is the value we added to cookies, check for error which is a boolean and message.
   		/* add data to locals too for accessing data from client */
   		throw redirect(307, '/dashboard');
   	}
   };
   ```

3. To get session data for the cookie

   ```ts
   /* hooks.server.ts */
   import sessionManager from 'sessionManagerFile';

   export const handle: Handle = async ({ event, resolve }) => {
   	/* your validation logic */
   	const { data, error, message } = sessionManager.getSession(event.cookies);
   	// data is the User session data we saved to redis while login, check for error which is a boolean and message.
   	/* do error check and then set data to locals as per your logic */
   };
   ```

4. To update session expiry in redis and cookies
   ```ts
   // in any server side file or endpoint where you can access browser cookies
   import sessionManager from 'sessionManagerFile';
   const { data, error, message } = await sessionManager.updateSessionExpiry(cookies);
   // data is going to be null or key which is updated, error is a boolean value and message a string
   ```
5. To delete session from redis and cookie from browser

   ```ts
   // Example it's a +page.server.ts

   import sessionManager from 'sessionManagerFile';
   export const actions: Actions = {
   	logout: async ({ cookies, locals }) => {
   		const { data, error, message } = await sessionManager.delSession(cookies);
   		// data is the value we deleted key, check for error which is a boolean and message.
   		/* clear your locals too */
   		throw redirect(307, '/login');
   	}
   };
   ```

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- USAGE EXAMPLES -->

## Usage

Examples are going to be added soon.

_For more examples, please refer to the [Examples](https://github.com/etherCorps/SK-Redis-SessionManager/tree/master/examples)_

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- ROADMAP -->

## Roadmap

See the [open issues](https://github.com/etherCorps/SK-Redis-SessionManager/issues) for a full list of proposed
features (and known issues).

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- CONTRIBUTING -->

[//]: # '## Contributing'

[//]: #

[//]: # 'Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any'

[//]: # 'contributions you make are **greatly appreciated**.'

[//]: #

[//]: # 'If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also'

[//]: # 'simply open an issue with the tag "enhancement".'

[//]: # "Don't forget to give the project a star! Thanks again!"

[//]: #

[//]: # '1. Fork the Project'

[//]: # '2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)'

[//]: # "3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)"

[//]: # '4. Push to the Branch (`git push origin feature/AmazingFeature`)'

[//]: # '5. Open a Pull Request'

[//]: #

[//]: # '<p align="right">(<a href="#readme-top">back to top</a>)</p>'

<!-- LICENSE -->

## License

Distributed under the MIT License. See `LICENSE.txt` for more information.

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- CONTACT -->

## Contact

Your Name - [@theether0](https://twitter.com/theether0) - meenashivam9650@gmail.com

Project
Link: [https://github.com/etherCorps/SK-Redis-SessionManager](https://github.com/etherCorps/SK-Redis-SessionManager)

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- ACKNOWLEDGMENTS -->

## Acknowledgments

- [logos-by-larkef from landingfolio](https://www.landingfolio.com/logos-by-larkef) :: For logo
- [connect-redis by TJ Holowaychuk](https://github.com/tj/connect-redis/tree/master) :: Creator of connect redis for express session.
- []()

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->

[contributors-shield]: https://img.shields.io/github/contributors/etherCorps/SK-Redis-SessionManager.svg?style=for-the-badge
[contributors-url]: https://github.com/etherCorps/SK-Redis-SessionManager/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/etherCorps/SK-Redis-SessionManager.svg?style=for-the-badge
[forks-url]: https://github.com/etherCorps/SK-Redis-SessionManager/network/members
[stars-shield]: https://img.shields.io/github/stars/etherCorps/SK-Redis-SessionManager.svg?style=for-the-badge
[stars-url]: https://github.com/etherCorps/SK-Redis-SessionManager/stargazers
[issues-shield]: https://img.shields.io/github/issues/etherCorps/SK-Redis-SessionManager.svg?style=for-the-badge
[issues-url]: https://github.com/etherCorps/SK-Redis-SessionManager/issues
[license-shield]: https://img.shields.io/github/license/etherCorps/SK-Redis-SessionManager.svg?style=for-the-badge
[license-url]: https://github.com/etherCorps/SK-Redis-SessionManager/blob/master/LICENSE.txt
[linkedin-shield]: https://img.shields.io/badge/-LinkedIn-black.svg?style=for-the-badge&logo=linkedin&colorB=555
[linkedin-url]: https://linkedin.com/in/theether0

[product-screenshot]: static/screenshot.png

[SvelteKit]: https://img.shields.io/badge/sveltekit-000000?style=for-the-badge&logo=svelte&logoColor=white

[SvelteKit-url]: https://kit.svelte.dev
[Redis]: https://img.shields.io/badge/Redis-DD0031?style=for-the-badge&logo=Redis&logoColor=white
[Redis-url]: https://redis.io/
[Svelte.dev]: https://img.shields.io/badge/Svelte-4A4A55?style=for-the-badge&logo=svelte&logoColor=FF3E00
[Svelte-url]: https://svelte.dev/


CHANGELOG.md
[//]: # '# Changelog'
[//]: #
[//]: # '## v1.0.0'
[//]: #
[//]: # '### Added or Changed'
[//]: # '- Added this changelog :)'
[//]: # '- Fixed typos in both templates'
[//]: # '- Back to top links'
[//]: # '- Added more "Built With" frameworks/libraries'
[//]: # '- Changed table of contents to start collapsed'
[//]: # '- Added checkboxes for major features on roadmap'
[//]: #
[//]: # '### Removed'
[//]: #
[//]: # '- Some packages/libraries from acknowledgements I no longer use'


src/lib/sessionManager.ts
import type * as ioRedis from 'ioredis';
import type { Cookies } from '@sveltejs/kit';
import { dev } from '$app/environment';
import crypto from 'crypto';

const defaultExpireSeconds: number = 60 * 60 * 24; // One day in seconds
const defaultRenewBeforeSeconds: number = 30 * 60; // 30 minutes in seconds
const defaultCookiesOption: CookieSerializeOptions = {
	path: '/',
	httpOnly: true,
	sameSite: 'strict',
	secure: !dev,
	maxAge: defaultExpireSeconds
};
export class RedisSessionStore {
	private readonly redisClient: RedisClientTypes;
	private readonly secret: string;
	private readonly cookieName: string;
	private readonly uniqueIdGenerator = crypto.randomUUID;
	private readonly prefix: string;
	private readonly signedCookies: boolean;
	private readonly encryptedCookies: boolean;
	private readonly useTTL: boolean;
	private readonly ttlSeconds: number | undefined;
	private readonly renewSessionBeforeExpire: boolean;
	private readonly renewBeforeSeconds: number;
	private readonly scanCount: number;
	private readonly serializer: Serializer;
	private readonly cookieOptions: CookieSerializeOptions;
	private readonly aesAlgorithm = 'aes-256-cbc';
	private readonly aesKey = Buffer.alloc(32);
	private readonly aesIv = crypto.randomBytes(16);
	constructor(options: redisSessionOptions) {
		if (!options.redisClient) {
			throw new Error('A client must be directly provided to the RedisStore');
		}

		options.redisClient.on('connect', () => {
			console.log('Connected to Redis');
		});

		options.redisClient.on('error', (error) => {
			console.error(`Error connecting to Redis: ${error}`);
			throw new Error('Unable to connect with RedisClient');
		});
		if (
			options.cookiesOptions &&
			options.cookiesOptions.maxAge &&
			options.cookiesOptions.maxAge < 1
		) {
			console.log('Please define a valid time in cookies maxAge parameter');
			throw new Error('Invalid maxAge in cookies options');
		}

		if (
			options.renewSessionBeforeExpire &&
			options.renewBeforeSeconds &&
			options.renewBeforeSeconds < 1
		) {
			console.log('Please define a valid time in renewBeforeSeconds');
			throw new Error('Invalid renewBeforeSeconds in options');
		}
		this.redisClient = options.redisClient;
		this.secret = options.secret;
		this.cookieName = options.cookieName || 'session';
		this.uniqueIdGenerator = crypto.randomUUID;
		this.prefix = options.prefix || 'sk-session:';
		this.signedCookies = options.signed || true;
		this.encryptedCookies = options.encrypted || false;
		this.useTTL = options.useTTL || true;
		this.renewSessionBeforeExpire = options.renewSessionBeforeExpire || false;
		this.renewBeforeSeconds = options.renewBeforeSeconds || defaultRenewBeforeSeconds;
		this.scanCount = Number( options.scanCount ) || 100;
		this.serializer = options.serializer || JSON;
		this.cookieOptions = { ...defaultCookiesOption, ...options.cookiesOptions };
		this.ttlSeconds = this.cookieOptions.maxAge;
		this.aesKey.write( options.secret );
	}

	async getSession(
		cookies: Cookies
	): Promise<{ data: any; error: boolean; message: string }> {
		let { data, error, message } = await this._validateCookie( cookies );
		if (error) return this._returnValid( data, error, message );
		const sessionData = await this.redisClient.get( `${this.prefix}${data}` );
		if (!sessionData) return this._returnValid( null, true, "Invalid session found." );
		let parsedSession;
		try {
			parsedSession = this.serializer.parse( sessionData );
		} catch (err) {
			console.log( err );
			return this._returnValid( null, true, "Unable to parse the session data." );
		}
		// logic for renew cookies and session before expire
		if (this.renewSessionBeforeExpire) {
			const sessionValidity = await this.redisClient.ttl( `${this.prefix}${data}` );
			if (sessionValidity < this.renewBeforeSeconds && this.ttlSeconds) {
				const { error, message } = await this.updateSessionExpiry( cookies, true, data );
				if (error) {
					console.log( message );
				}
			}
		}
		return this._returnValid( parsedSession, false, "Session Data" ); // return session data
	}

	async createNewSession(
		cookies: Cookies,
		sessionData = {},
		key?: string
	): Promise<{ data: any; error: boolean; message: string }> {
		let serializedSessionData;
		try {
			serializedSessionData = this.serializer.stringify( sessionData );
		} catch (er) {
			console.log( "Error in Set Session while serializing", er );
			return this._returnValid( null, true, "Unable to stringify session data." );
		}
		const uniqueKey = key || this.uniqueIdGenerator();
		if (typeof uniqueKey !== "string")
			return this._returnValid(
				null,
				true,
				'Please check your key is a string or uniqueIdGenerator return type'
			);
		const keyWithPrefix = this.prefix + uniqueKey;
		const args = [keyWithPrefix, serializedSessionData];
		if (this.useTTL && this.ttlSeconds) {
			args.push( "EX", this.ttlSeconds );
		}
		// @ts-ignore
		this.redisClient.set( args );
		let finalKey = uniqueKey;
		if (this.signedCookies) finalKey = await this._signKey( finalKey );
		if (this.encryptedCookies) finalKey = await this._encrypt( finalKey );
		cookies.set( this.cookieName, finalKey, this.cookieOptions );
		return this._returnValid( { uniqueKey, finalKey }, false, "Ready get set go" ); // Returns cookie value after setting to cookie
	}

	async updateSessionExpiry(
		cookies: Cookies,
		skipValidation: boolean = false,
		key: string = ""
	): Promise<{ data: any; error: boolean; message: string }> {
		let uniqueKey = key;
		if (!skipValidation) {
			const { data, error, message } = await this._validateCookie( cookies );
			if (error) {
				console.log( "Error in updateSessionExpiry method", message );
				return this._returnValid( data, error, "Unable to validate key while updating session" );
			}
			uniqueKey = data;
		}
		let isExpireTimeUpdated = 1;
		if (this.useTTL && this.ttlSeconds) {
			isExpireTimeUpdated = await this.redisClient.expire(
				`${this.prefix}${uniqueKey}`,
				this.ttlSeconds as number
			);
		}
		if (isExpireTimeUpdated) {
			let finalKey = uniqueKey;
			if (this.signedCookies) finalKey = await this._signKey( finalKey );
			if (this.encryptedCookies) finalKey = await this._encrypt( finalKey );
			cookies.set( this.cookieName, finalKey, this.cookieOptions );
			return this._returnValid(
				{ uniqueKey, finalKey },
				false,
				"Session validity extended successfully"
			); // return cookie value after updating expiry
		}
		return this._returnValid( null, true, "Unable to extended session validity" );
	}

	async delSession(
		cookies: Cookies
	): Promise<{ data: any; error: boolean; message: string }> {
		const { data, error, message } = await this._validateCookie( cookies );
		if (error) {
			console.log( "Error in delSession method", message );
			return this._returnValid( data, error, "Unable to validate key while deleting" );
		}
		const deleteData = await this.redisClient.del( `${this.prefix}${data}` );
		if (!deleteData) return this._returnValid( null, true, `Key not found while deleting` );
		await this.deleteCookie( cookies );
		return this._returnValid( data, false, `Key successfully deleted` ); // Returns unique key without prefix which is deleted from redis
	}

	async deleteCookie(cookies: Cookies) {
		const allCookieOptionsCopy = { ...this.cookieOptions };
		delete allCookieOptionsCopy.maxAge;
		try {
			cookies.delete( this.cookieName, allCookieOptionsCopy );
		} catch (err) {
			console.log( "error while deleting cookies in deleteCookie method", err );
		}
	}

	async _validateCookie(cookies: Cookies) {
		const cookiesSessionKey = cookies.get( this.cookieName );
		if (!cookiesSessionKey) return this._returnValid( null, true, "No session found in cookies." );
		let verifiedSessionKey = cookiesSessionKey;
		if (this.encryptedCookies) {
			const { encrypted, iv } = this.serializer.parse( verifiedSessionKey );
			verifiedSessionKey = (await this._decrypt( { encrypted, iv } )) as string;
		}
		if (this.signedCookies)
			verifiedSessionKey = (await this._verifyKeySignature( verifiedSessionKey )) as string;
		if (!verifiedSessionKey)
			return this._returnValid( null, true, "Cookies session is not verified." );
		return this._returnValid( verifiedSessionKey, false, "Successfully validated cookies" ); // it returns uniques that will make redis key with prefix
	}
	_signKey = async (key: string) => {
		const newDigest = await crypto.createHmac('sha256', this.secret).update(key).digest('hex');
		return `${key}.${newDigest}`;
	};

	_encrypt = async (keyToBeEncrypted: string) => {
		const cipherAES = crypto.createCipheriv( this.aesAlgorithm, this.aesKey, this.aesIv );
		let encrypted;
		try {
			encrypted = cipherAES.update( keyToBeEncrypted, "utf8", "hex" );
			encrypted += cipherAES.final( "hex" );
		} catch (e) {
			console.log( "Error in encrypt method", e );
			throw new Error( "Encryption Error" );
		}
		const encryptedCookiesValue = this.serializer.stringify( {
			encrypted,
			iv: this.aesIv.toString( "hex" )
		} );
		return encryptedCookiesValue;
	};

	_decrypt = async ({ encrypted, iv }: { encrypted: string; iv: string }) => {
		try {
			const decipher = crypto.createDecipheriv(
				this.aesAlgorithm,
				this.aesKey,
				Buffer.from( iv, "hex" )
			);
			let decrypted = decipher.update( encrypted, "hex", "utf8" );
			decrypted += decipher.final( "utf8" );
			return decrypted as string;
		} catch (e) {
			console.log( "decryption error: ", e );
			return null;
		}
	};

	_verifyKeySignature = async (signedCookie: string) => {
		const valueWithSignature = signedCookie.split( "." );
		try {
			const value = valueWithSignature[0];
			const signature = valueWithSignature[1];
			const hmac = crypto.createHmac( "sha256", this.secret );
			hmac.update( value );
			const expectedSignature = hmac.digest( "hex" );
			const isValidSignature = crypto.timingSafeEqual(
				Buffer.from(signature, 'hex'),
				Buffer.from(expectedSignature, 'hex')
			);
			if (!isValidSignature) {
				return null;
			}
			return value;
		} catch (e) {}
	};
	_returnValid(data: any, error: boolean, message: string) {
		return { data, error, message };
	}

	// clear(cb = noop) {
	//   this._getAllKeys((err, keys) => {
	//     if (err) return cb(err)
	//     this.client.del(keys, cb)
	//   })
	// }
	//
	// length(cb = noop) {
	//   this._getAllKeys((err, keys) => {
	//     if (err) return cb(err)
	//     return cb(null, keys.length)
	//   })
	// }
	//
	// ids(cb = noop) {
	//   let prefixLen = this.prefix.length;
	//
	//   this._getAllKeys((err, keys) => {
	//     if (err) return cb(err)
	//     keys = keys.map((key) => key.substr(prefixLen))
	//     return cb(null, keys)
	//   })
	// }
	//
	// all(cb = noop) {
	//   let prefixLen = this.prefix.length
	//
	//   this._getAllKeys((err, keys) => {
	//     if (err) return cb(err)
	//     if (keys.length === 0) return cb(null, [])
	//
	//     this.client.mget(keys, (err, sessions) => {
	//       if (err) return cb(err)
	//
	//       let result
	//       try {
	//         result = sessions.reduce((accum, data, index) => {
	//           if (!data) return accum
	//           data = this.serializer.parse(data)
	//           data.id = keys[index].substr(prefixLen)
	//           accum.push(data)
	//           return accum
	//         }, [])
	//       } catch (e) {
	//         err = e
	//       }
	//       return cb(err, result)
	//     })
	//   })
	// }
	//
	//
	// _getAllKeys(cb = noop) {
	//   let pattern = this.prefix + "*"
	//   this._scanKeys({}, 0, pattern, this.scanCount, cb)
	// }
	//
	// _scanKeys(keys = {}, cursor, pattern, count, cb = noop) {
	//   let args = [cursor, "match", pattern, "count", count]
	//   this.client.scan(args, (err, data) => {
	//     if (err) return cb(err)
	//
	//     let [nextCursorId, scanKeys] = data
	//     for (let key of scanKeys) {
	//       keys[key] = true
	//     }
	//
	//     // This can be a string or a number. We check both.
	//     if (Number(nextCursorId) !== 0) {
	//       return this._scanKeys(keys, nextCursorId, pattern, count, cb)
	//     }
	//
	//     cb(null, Object.keys(keys))
	//   })
	// }
}

export type RedisClientTypes = ioRedis.Redis | ioRedis.Cluster;
export interface redisSessionOptions {
	redisClient: RedisClientTypes;
	secret: string;
	cookieName?: string;
	prefix?: string;
	signed?: boolean;
	encrypted?: boolean;
	useTTL?: boolean;
	renewSessionBeforeExpire?: boolean;
	renewBeforeSeconds?: number;
	scanCount?: number; // currently not using, future plans to add some functionality
	serializer?: Serializer | JSON;
	cookiesOptions?: CookieSerializeOptions;
}

export type Serializer = {
	stringify: Function;
	parse: Function;
};

export interface CookieSerializeOptions {
	/**
	 * Specifies the value for the {@link https://tools.ietf.org/html/rfc6265#section-5.2.3|Domain Set-Cookie attribute}. By default, no
	 * domain is set, and most clients will consider the cookie to apply to only
	 * the current domain.
	 */
	domain?: string | undefined;

	/**
	 * Specifies a function that will be used to encode a cookie's value. Since
	 * value of a cookie has a limited character set (and must be a simple
	 * string), this function can be used to encode a value into a string suited
	 * for a cookie's value.
	 *
	 * The default function is the global `encodeURIComponent`, which will
	 * encode a JavaScript string into UTF-8 byte sequences and then URL-encode
	 * any that fall outside the cookie range.
	 */
	encode?(value: string): string;

	/**
	 * Specifies the `Date` object to be the value for the {@link https://tools.ietf.org/html/rfc6265#section-5.2.1|`Expires` `Set-Cookie` attribute}. By default,
	 * no expiration is set, and most clients will consider this a "non-persistent cookie" and will delete
	 * it on a condition like exiting a web browser application.
	 *
	 * *Note* the {@link https://tools.ietf.org/html/rfc6265#section-5.3|cookie storage model specification}
	 * states that if both `expires` and `maxAge` are set, then `maxAge` takes precedence, but it is
	 * possible not all clients by obey this, so if both are set, they should
	 * point to the same date and time.
	 */
	expires?: Date | undefined;
	/**
	 * Specifies the boolean value for the {@link https://tools.ietf.org/html/rfc6265#section-5.2.6|`HttpOnly` `Set-Cookie` attribute}.
	 * When truthy, the `HttpOnly` attribute is set, otherwise it is not. By
	 * default, the `HttpOnly` attribute is not set.
	 *
	 * *Note* be careful when setting this to true, as compliant clients will
	 * not allow client-side JavaScript to see the cookie in `document.cookie`.
	 */
	httpOnly?: boolean | undefined;
	/**
	 * Specifies the number (in seconds) to be the value for the `Max-Age`
	 * `Set-Cookie` attribute. The given number will be converted to an integer
	 * by rounding down. By default, no maximum age is set.
	 *
	 * *Note* the {@link https://tools.ietf.org/html/rfc6265#section-5.3|cookie storage model specification}
	 * states that if both `expires` and `maxAge` are set, then `maxAge` takes precedence, but it is
	 * possible not all clients by obey this, so if both are set, they should
	 * point to the same date and time.
	 */
	maxAge?: number | undefined;
	/**
	 * Specifies the value for the {@link https://tools.ietf.org/html/rfc6265#section-5.2.4|`Path` `Set-Cookie` attribute}.
	 * By default, the path is considered the "default path".
	 */
	path?: string | undefined;
	/**
	 * Specifies the `string` to be the value for the [`Priority` `Set-Cookie` attribute][rfc-west-cookie-priority-00-4.1].
	 *
	 * - `'low'` will set the `Priority` attribute to `Low`.
	 * - `'medium'` will set the `Priority` attribute to `Medium`, the default priority when not set.
	 * - `'high'` will set the `Priority` attribute to `High`.
	 *
	 * More information about the different priority levels can be found in
	 * [the specification][rfc-west-cookie-priority-00-4.1].
	 *
	 * **note** This is an attribute that has not yet been fully standardized, and may change in the future.
	 * This also means many clients may ignore this attribute until they understand it.
	 */
	priority?: 'low' | 'medium' | 'high' | undefined;
	/**
	 * Specifies the boolean or string to be the value for the {@link https://tools.ietf.org/html/draft-ietf-httpbis-rfc6265bis-03#section-4.1.2.7|`SameSite` `Set-Cookie` attribute}.
	 *
	 * - `true` will set the `SameSite` attribute to `Strict` for strict same
	 * site enforcement.
	 * - `false` will not set the `SameSite` attribute.
	 * - `'lax'` will set the `SameSite` attribute to Lax for lax same site
	 * enforcement.
	 * - `'strict'` will set the `SameSite` attribute to Strict for strict same
	 * site enforcement.
	 *  - `'none'` will set the SameSite attribute to None for an explicit
	 *  cross-site cookie.
	 *
	 * More information about the different enforcement levels can be found in {@link https://tools.ietf.org/html/draft-ietf-httpbis-rfc6265bis-03#section-4.1.2.7|the specification}.
	 *
	 * *note* This is an attribute that has not yet been fully standardized, and may change in the future. This also means many clients may ignore this attribute until they understand it.
	 */
	sameSite?: true | false | 'lax' | 'strict' | 'none' | undefined;
	/**
	 * Specifies the boolean value for the {@link https://tools.ietf.org/html/rfc6265#section-5.2.5|`Secure` `Set-Cookie` attribute}. When truthy, the
	 * `Secure` attribute is set, otherwise it is not. By default, the `Secure` attribute is not set.
	 *
	 * *Note* be careful when setting this to `true`, as compliant clients will
	 * not send the cookie back to the server in the future if the browser does
	 * not have an HTTPS connection.
	 */
	secure?: boolean | undefined;
}
